pkg/lib/synthesize/generate.go

```diff
@@ -0,0 +1,174 @@
+	bundleDir, err := filepath.Abs(bundleDir)
+	if err != nil {
+		return err
+	}
+
+	// find annotations.yaml file in the bundle dir.
+	var annotations bundle.Annotations
+	var annotationsFilePath string
+	if err := filepath.Walk(bundleDir, func(path string, info os.FileInfo, err error) error {
+		if annotations.IsComplete() {
+			return nil
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		annotationsFilePath = path
+		content, err := ioutil.ReadFile(annotationsFilePath)
+		if err != nil {
+			return err
+		}
+
+		// tries to parse the content as annotations else go to the next file.
+		annotations, _ = bundle.GetAnnotations(content)
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if !annotations.IsComplete() {
+		return fmt.Errorf("failed to find annotations YAML file in the bundle")
+	}
+
+	s, err := SynthesizeCsvFromDirectories(filepath.Join(bundleDir, annotations.ManifestsDirName),
+		filepath.Join(bundleDir, annotations.MetadataDirName))
+	if err != nil {
+		return err
+	}
+
+	csv, err := s.GenerateCSV()
+	if err != nil {
+		return err
+	}
+
+	// generate CSV file to the outputDir.
+	manifestsDir := path.Join(bundleDir, annotations.ManifestsDirName)
+	if _, err := os.Stat(manifestsDir); os.IsNotExist(err) {
+		err = os.MkdirAll(manifestsDir, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	filePath := path.Join(manifestsDir, SynthesizedCSVFileName)
+
+	err = WriteStructToFile(csv, filePath)
+	if err != nil {
+		return fmt.Errorf("error writing CSV file, %v", err)
+	}
+
+	if s.existingCSVPath != "" {
+		log.Warnf("Overwriting an existing ClusterServiceVersion in the bundle at %s", s.existingCSVPath)
+		if s.existingCSVPath != filePath {
+			os.Remove(s.existingCSVPath)
+		}
+	}
+
+	// Update media type after successfully creating the CSV
+	err = os.Remove(annotationsFilePath)
+	if err != nil {
+		return fmt.Errorf("error deleting annotations YAML file %s, %v", annotationsFilePath, err)
+	}
+
+	annotations.MediaType = bundle.RegistryV1Type
+	annotationsContent, err := bundle.GenerateAnnotations(annotations)
+	err = bundle.WriteFile(bundle.AnnotationsFile, path.Join(bundleDir, annotations.MetadataDirName), annotationsContent)
+	if err != nil {
+		return err
+	}
+
+	// Validate the bundle directory
+	logger := log.NewEntry(log.New())
+	log.SetLevel(log.DebugLevel)
+
+	bundleValidator := bundle.NewBundleValidator(logger)
+	if err := bundleValidator.ValidateBundleFormat(bundleDir); err != nil {
+		return err
+	}
+
+	if err := bundleValidator.ValidateBundleContent(manifestsDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type MarshalFunc func(interface{}) ([]byte, error)
+
+// WriteStructToFile writes a struct's content to a YAML file.
+// This strategy is borrowed from SDK to marshall an object with m and removes runtime-managed fields:
+// 'status', 'creationTimestamp'
+// https://github.com/operator-framework/operator-sdk/blob/master/internal/util/k8sutil/object.go#L25
```

Comment:
Doc tags like this that just point to master are going to be prone to get out of date. I think we should either point this link to a specific commit or just remove it entirely.